src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java

```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+
+        // null header allows any ref name set
+        final String unresolvableRefName = "unresolvable";
+        sam.setReferenceName(unresolvableRefName);
+        Assert.assertTrue(unresolvableRefName.equals(sam.getReferenceName()));
+    }
+
+    @Test
+    public void testNullHeaderReferenceIndex() {
+        SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        final SAMFileHeader header = sam.getHeader();
+
+        // NO_ALIGNMENT_REFERENCE
+        sam.setHeader(null);
+        sam.setReferenceIndex(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+        Assert.assertTrue(sam.getReferenceIndex().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
+        Assert.assertTrue(sam.getReferenceName().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME));
+
+        // setting the header to null allows reference name to be reset
+        sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        sam.setHeader(null);
+        sam.setReferenceName(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME);
+        Assert.assertTrue(sam.getReferenceIndex() == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+
+        // force the internal SAMRecord reference index value to (null) initial state
+        sam = new SAMRecord(null);
+        Assert.assertTrue(sam.getReferenceIndex() == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+
+        sam = new SAMRecord(null);
+        final String unresolvableRefName = "unresolvable";
+        sam.setReferenceName(unresolvableRefName);
+        // now force the SAMRecord to try to resolve the unresolvable name
+        sam.setHeader(header);
+        Assert.assertTrue(sam.getReferenceIndex() == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+    }
+
+    @Test(expectedExceptions=IllegalArgumentException.class)
+    public void testNullHeaderSetReferenceIndex() {
+        final SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        sam.setHeader(null);
+        // setReferenceIndex with null header throws
+        sam.setReferenceIndex(3);
+    }
+
+    @Test(expectedExceptions=IllegalArgumentException.class)
+    public void testNullHeaderForceIndexResolutionFailure() {
+        // force the internal SAMRecord reference index value to null initial state
+        final SAMRecord sam = new SAMRecord(null);
+        sam.setReferenceName("unresolvable");
+        sam.getReferenceIndex();
     }
+
+    @Test
+    public void testNullHeaderMateReferenceName() {
+        SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        final SAMFileHeader samHeader = sam.getHeader();
+        sam.setMateReferenceName("chr1");
+        final String origMateRefName = sam.getMateReferenceName();
+        final int originalMateRefIndex = sam.getMateReferenceIndex();
+
+        // setting header to null retains ref name and index
```

Comment:
It shouldn't retain the index, it should null it out (and you should check that it does, since `mMateReferenceIndex` has protected access and you're in the right package).